File: Grunt/Grunt/Endpoints/HaloCoreEndpoints.cs
// <copyright file="HaloCoreEndpoints.cs" company="Den Delimarsky">
// Developed by Den Delimarsky.
// Den Delimarsky licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
// The underlying API powering Grunt is managed by 343 Industries and Microsoft. This wrapper is not endorsed by 343 Industries or Microsoft.
// </copyright>

namespace OpenSpartan.Grunt.Endpoints
{
    /// <summary>
    /// Container for Halo API endpoints.
    /// </summary>
    internal class HaloCoreEndpoints
    {
        /// <summary>
        /// Relying party for use with the Xbox Live authentication flow, associated with the Halo Waypoint service.
        /// </summary>
        internal static readonly string HaloWaypointXstsRelyingParty = "https://prod.xsts.halowaypoint.com/";

        // ====================
        // Origins
        // ====================

        /// <summary>
        /// Game CMS origin.
        /// </summary>
        internal static readonly string GameCmsOrigin = "gamecms-hacs";

        /// <summary>
        /// Economy origin.
        /// </summary>
        internal static readonly string EconomyOrigin = "economy";

        /// <summary>
        /// Authoring origin.
        /// </summary>
        internal static readonly string AuthoringOrigin = "authoring-infiniteugc";

        /// <summary>
        /// Discovery origin.
        /// </summary>
        internal static readonly string DiscoveryOrigin = "discovery-infiniteugc";

        /// <summary>
        /// Halo Infinite lobby origin.
        /// </summary>
        internal static readonly string HaloInfiniteLobbyOrigin = "lobby-hi";

        /// <summary>
        /// Settings origin.
        /// </summary>
        internal static readonly string SettingsOrigin = "settings";

        /// <summary>
        /// Skill origin.
        /// </summary>
        internal static readonly string SkillOrigin = "skill";

        /// <summary>
        /// Ban processor origin.
        /// </summary>
        internal static readonly string BanProcessorOrigin = "banprocessor";

        /// <summary>
        /// Stats origin.
        /// </summary>
        internal static readonly string StatsOrigin = "halostats";

        /// <summary>
        /// Text origin.
        /// </summary>
        internal static readonly string TextOrigin = "text";

        /// <summary>
        /// Content HACS origin.
        /// </summary>
        internal static readonly string ContentHacsOrigin = "content-hacs";

        // ====================
        // Service domains
        // ====================

        /// <summary>
        /// Halo Waypoint service domain used for all Halo API calls.
        /// </summary>
        internal static readonly string ServiceDomain = "svc.halowaypoint.com:443";

        // ====================
        // Service endpoints
        // ====================

        /// <summary>
        /// Endpoint used to produce the Spartan token.
        /// </summary>
        internal static readonly string SpartanTokenEndpoint = "https://settings.svc.halowaypoint.com/spartan-token";

        /// <summary>
        /// Endpoint used to obtain the Halo Infinite endpoints.
        /// </summary>
        internal static readonly string HaloInfiniteEndpointsEndpoint = "https://settings.svc.halowaypoint.com/settings/hipc/e2a0a7c6-6efe-42af-9283-c2ab73250c48";

        /// <summary>
        /// Endpoint used to obtain the Halo 5 endpoints.
        /// </summary>
        internal static readonly string Halo5EndpointsEndpoint = "https://settings.svc.halowaypoint.com/settings/h5pc/a1b344c4-91a3-47f7-92f4-95784cda3cd2";
    }
}

File: src/node/src/endpoints/halo-core-endpoints.ts
/**
 * Halo API endpoint configuration constants.
 *
 * These define the service origins for different API modules.
 * The Halo Infinite API is distributed across multiple services,
 * each handling a specific domain of functionality.
 */
export const HALO_CORE_ENDPOINTS = {
  /**
   * Base service domain for all Halo Waypoint services.
   */
  SERVICE_DOMAIN: 'svc.halowaypoint.com',

  // ─────────────────────────────────────────────────────────────────
  // Service Origins (prepended to SERVICE_DOMAIN)
  // ─────────────────────────────────────────────────────────────────

  /** Game CMS for static content (challenges, items, etc.) */
  GAME_CMS_ORIGIN: 'gamecms-hacs',

  /** Economy service for stores, inventory, customization */
  ECONOMY_ORIGIN: 'economy',

  /** UGC authoring service for creating/editing user content */
  AUTHORING_ORIGIN: 'authoring-infiniteugc',

  /** UGC discovery service for searching user content */
  DISCOVERY_ORIGIN: 'discovery-infiniteugc',

  /** Lobby service for multiplayer lobbies and presence */
  LOBBY_ORIGIN: 'lobby-hi',

  /** Settings and configuration service */
  SETTINGS_ORIGIN: 'settings',

  /** Skill and CSR rating service */
  SKILL_ORIGIN: 'skill',

  /** Ban processor service */
  BAN_PROCESSOR_ORIGIN: 'banprocessor',

  /** Stats service for match history and service records */
  STATS_ORIGIN: 'halostats',

  /** Text moderation service */
  TEXT_ORIGIN: 'text',

  /** Content service for articles and news */
  CONTENT_ORIGIN: 'content-hacs',

  // ─────────────────────────────────────────────────────────────────
  // Authentication Endpoints
  // ─────────────────────────────────────────────────────────────────

  /**
   * Endpoint for obtaining a Spartan token from an XSTS token.
   */
  SPARTAN_TOKEN_ENDPOINT: 'https://settings.svc.halowaypoint.com/spartan-token',

  /**
   * Endpoint for discovering available Halo Infinite API endpoints.
   * Returns configuration for all available services.
   */
  HALO_INFINITE_SETTINGS:
    'https://settings.svc.halowaypoint.com/settings/hipc/e2a0a7c6-6efe-42af-9283-c2ab73250c48',

  // ─────────────────────────────────────────────────────────────────
  // Xbox Live / XSTS Configuration
  // ─────────────────────────────────────────────────────────────────

  /**
   * Relying party URL for XSTS token exchange.
   * Use this when requesting an XSTS token for Halo Waypoint.
   */
  HALO_WAYPOINT_XSTS_RELYING_PARTY: 'https://prod.xsts.halowaypoint.com/',

  // ─────────────────────────────────────────────────────────────────
  // Blob Storage
  // ─────────────────────────────────────────────────────────────────

  /**
   * Base URL for UGC blob storage (maps, game variants, etc.)
   */
  BLOBS_ORIGIN: 'blobs-infiniteugc',
} as const;

/**
 * Waypoint-specific endpoints for non-game services.
 */
export const WAYPOINT_ENDPOINTS = {
  /**
   * Base domain for Halo Waypoint web services.
   */
  WEB_DOMAIN: 'www.halowaypoint.com',

  /**
   * API subdomain for Waypoint services.
   */
  API_DOMAIN: 'api.halowaypoint.com',

  /**
   * Profile API origin.
   */
  PROFILE_ORIGIN: 'profile',

  /**
   * Redemption API origin for code redemption.
   */
  REDEMPTION_ORIGIN: 'redemption',
} as const;

/**
 * Build a full URL from an origin and path.
 *
 * @param origin - The service origin (e.g., 'halostats')
 * @param path - The API path (e.g., '/hi/players/xuid(...)/matches')
 * @returns Full HTTPS URL
 *
 * @example
 * ```typescript
 * const url = buildServiceUrl(HALO_CORE_ENDPOINTS.STATS_ORIGIN, '/hi/players/xuid(123)/matches');
 * // Returns: 'https://halostats.svc.halowaypoint.com/hi/players/xuid(123)/matches'
 * ```
 */
export function buildServiceUrl(origin: string, path: string): string {
  return `https://${origin}.${HALO_CORE_ENDPOINTS.SERVICE_DOMAIN}${path}`;
}
